flask/launcher/src/main/java/net/corda/flask/launcher/JavaProcessBuilder.java

```diff
@@ -0,0 +1,167 @@
+                    case '\\':
+                        sb.append("\\\\");
+                        break;
+                    default:
+                        sb.append(c);
+                        break;
+                }
+            }
+            sb.append('"');
+            if(++i < strings.size()) {
+                sb.append(' ');
+            }
+        }
+        return sb.toString();
+    }
+
+    @SneakyThrows
+    public int exec() {
+        Process process = build().inheritIO().start();
+        try {
+            return process.waitFor();
+        } finally {
+            if(process.isAlive()) {
+                process.destroy();
+            }
+            if (process.isAlive()) {
+                process.destroyForcibly();
+            }
+        }
+
+    }
+
+    @SneakyThrows
+    public ProcessBuilder build() {
+        ArrayList<String> cmd = new ArrayList<>();
+        Path javaBin = Paths.get(javaHome, "bin", "java");
+        cmd.add(javaBin.toString());
+        cmd.addAll(jvmArgs);
+        if(!classpath.isEmpty()) {
+            cmd.add("-cp");
+            cmd.add(classpath.stream().collect(Collectors.joining(System.getProperty("path.separator"))));
+        }
+        for(Map.Entry<Object, Object> entry : properties.entrySet()) {
+            cmd.add(String.format("-D%s=%s", entry.getKey(), entry.getValue()));
+        }
+        for(JavaAgent javaAgent : javaAgents) {
+            StringBuilder sb = new StringBuilder("-javaagent:" + javaAgent.jar.toString());
+            String agentArguments = javaAgent.args;
+            if(agentArguments != null) {
+                sb.append('=');
+                sb.append(agentArguments);
+            }
+            cmd.add(sb.toString());
+        }
+        if(executableJar != null) {
+            cmd.add("-jar");
+            cmd.add(executableJar.toString());
+        } else if(mainClassName != null) {
+            cmd.add(mainClassName);
+        } else {
+            throw new IllegalArgumentException(
+                    "Either a main class or the path to an executable jar file have to be specified");
+        }
+        cmd.addAll(cliArgs);
+
+        int cmdLength = 0;
+        for(String part : cmd) {
+            cmdLength += part.length();
+        }
+        //Add space between arguments
+        cmdLength += cmd.size() - 1;
+
+        log.debug("Spawning new process with command line: [{}]",
+                cmd.stream().map(s -> "\"" + s + "\"").collect(Collectors.joining(", ")));
+        if(cmdLength < 1024 || cmd.size() == 1) {
+            return new ProcessBuilder(cmd);
+        } else {
+            Path argumentFile = Files.createTempFile(PROCESS_BUILDER_PREFIX, ".arg");
+            Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
+                @Override
+                @SneakyThrows
+                public void run() {
+                    Files.delete(argumentFile);
+                }
+            }));
```

Comment:
all the `@SneakyThrows` annotation does is make `javac` happy about uncaught checked exceptions, which is exactly what happens every time you write a line of `Kotlin` or `Groovy` code, if you aren't concerned about not catching `IOException` in Groovy or Kotlin, you shouldn't be in Java either

Reply:
regarding the shutdown hook, this part needs to be reviewed and better tested because I'm also concerned about zombie process creation

Reply:
FYI I _don't_ like the way Kotlin ignores checked exceptions because it throws a spanner into Java / Kotlin interoperability, particularly for Java CorDapp developers.

A `Runnable` is not expected throw an `IOException`, but yours will if someone deletes that file first. So what happens then?

Reply:
my `Runnable` can throw an `IOException`, should that happen there is no way you can catch that exception elsewhere in the callstack (this is due to `javac` not allowing you to catch a checked exception on a `try` block that doesn't contain any code that is declared to throw that exception type). 

This is the main drawback of `@SneakyThrows` (but I have no intent of catching that `IOException` anywhere here).
To be more accurate, you can actually catch those exceptions if there is a try block up in the stack that already contains code that throws the same exception type, e.g.
```java
class Foo {
    @SneakyThrows
    void bar() {
        Files.delete(Paths.get("."));
    }
    
    //This method doesn't compile since "bar" isn't declared 
    //to throw IOException which is a checked exception
    void bar2() {
        try {
            bar();
        } catch (IOException e) {
        }
    }

    //This method compiles and catches IOException
    //thrown by both 'Files.delete' and 'bar'
    void bar3() {
        try {
            Files.delete(Paths.get("."));
            bar();
        } catch (IOException e) {
        }
    }
}
```

anyway, if a checked exception is thrown from a `@SneakyThrows` annotated method, the most common result is that the exception causes the process to terminate with a quite clear error message and stacktrace since the exception isn't wrapped in a `RuntimeException` as it is customary in Java for the sole purpose of making the code compile

Reply:
> my `Runnable` can throw an `IOException`, should that happen there is no way you can catch that exception elsewhere in the callstack
> This is the main drawback of `@SneakyThrows` (but I have no intent of catching that `IOException` anywhere here).

And there's your bug: This is a shutdown hook, and there may still be others that need to run after yours (not necessarily installed by you). They aren't _*ALLOWED*_ to let exceptions escape and so you _*MUST*_ catch `IOException` here to allow the shutdown processing to continue.

Reply:
shutdown hooks run in parallel, not sequentially, each one in a in different thread (note that `addShutdownHook` accepts a `Thread` as a parameter), so if one throws an uncaught exception that doesn't mean the others won't be invoked

you can see this with
```
#!/usr/bin/env groovy
Runtime runtime = Runtime.getRuntime()

runtime.addShutdownHook(new Thread({
  throw new RuntimeException('KA-BOOM!!')
}))

runtime.addShutdownHook(new Thread({
  println('here')
}))

```
the string "here" is always printed...

Reply:
the [java documentation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Runtime.html#addShutdownHook(java.lang.Thread)) also clearly states that an uncaught exception in a shutdown hook will simply cause its thread to terminate, not the JVM to exit